jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -503,130 +509,40 @@ public static String parentPath(String p)
      */
     public static String canonicalPath(String path)
     {
-        if (path==null || path.length()==0)
+        if (path == null || path.isEmpty() || !path.contains("."))
             return path;
 
-        int end=path.length();
-        int start = path.lastIndexOf('/', end);
+        if(path.startsWith("/.."))
+            return null;
 
-    search:
-        while (end>0)
-        {
-            switch(end-start)
-            {
-              case 2: // possible single dot
-                  if (path.charAt(start+1)!='.')
-                      break;
-                  break search;
-              case 3: // possible double dot
-                  if (path.charAt(start+1)!='.' || path.charAt(start+2)!='.')
-                      break;
-                  break search;
-            }
-            
-            end=start;
-            start=path.lastIndexOf('/',end-1);
-        }
+        List<String> directories = new LinkedList<>();
```

Comment:
What do you mean when you say is a garbage generator?? It's because removed elements will be on memory until the LinkedList is destroyed? 
If so, it would not me much problem due that as how you said these list are short. But if you feel more comfortable using the ArrayList I can change it.
I already tried with the Array and I am having some problems with the OutOfBoundException, also it can be fixed but I would have to add a bit more code. 
How do you like the most?